stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/io/channels/base/Channel.java

```diff
@@ -39,50 +43,154 @@
+            throw new BallerinaIOException(message);
+        }
+    }
 
-    public Channel(ByteChannel channel, int size) throws BallerinaIOException {
-        super(channel);
-        contentBuffer = new Buffer(size);
+    /**
+     * Will be used when performing direct transfer operations from OS cache.
+     *
+     * @param position   starting position of the bytes to be transferred.
+     * @param count      number of bytes to be transferred.
+     * @param dstChannel destination channel to transfer.
+     * @throws IOException during I/O error.
+     */
+    public abstract void transfer(int position, int count, WritableByteChannel dstChannel) throws IOException;
+
+    /**
+     * Specifies whether the channel has reached to it's end.
+     *
+     * @return true if the channel has reached to it's end
+     */
+    public boolean hasReachedEnd() {
+        return hasReachedToEnd;
     }
 
     /**
      * <p>
-     * Merge the list of buffers together into one.
+     * Async read bytes from the channel.
      * </p>
+     *
+     * @param buffer the buffer which will hold the content.
+     * @return the number of bytes read.
+     * @throws IOException errors occur during reading from channel.
+     */
+    public int read(ByteBuffer buffer) throws IOException {
+        int readBytes = reader.read(buffer, channel);
+        if (readBytes <= 0) {
+            //Since we're counting the bytes if a value < 0 is returned, this will be re-set
+            readBytes = 0;
+            hasReachedToEnd = true;
+        }
+        return readBytes;
+    }
+
+    /**
      * <p>
-     * <b>Note : </b> This will be used to consolidate buffers read in multiple iterations into one.
+     * Writes provided buffer content to the channel.
      * </p>
      *
-     * @param bufferList the list which contains all bytes read through the buffer.
-     * @param size       the size of all the bytes read.
-     * @return the response buffer which contains all consolidated bytes.
+     * @param content the buffer which holds the content.
+     * @return the number of bytes written to the channel.
+     * @throws IOException errors occur during writing data to channel.
+     */
+    public int write(ByteBuffer content) throws IOException {
+        return writer.write(content, channel);
+    }
+
+    /**
+     * This will return {@link InputStream} from underlying {@link ByteChannel}.
+     *
+     * @return An {@link InputStream}
+     * @throws BallerinaIOException error occur during obtaining input-stream.
      */
-    private byte[] merge(List<ByteBuffer> bufferList, int size) {
-        ByteBuffer consolidatedBuffer = ByteBuffer.allocate(size);
-        for (ByteBuffer buffer : bufferList) {
-            consolidatedBuffer.put(buffer);
+    public InputStream getInputStream() throws BallerinaIOException {
+        if (!channel.isOpen()) {
+            String message = "Channel is already closed.";
+            throw new BallerinaIOException(message);
```

Comment:
Shall we convert this to a single line?

Reply:
This way is more readable and extensible IMO. Let's say we want to log the message. If so instead of duplicating we could use the message itself